src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs

```diff
@@ -691,13 +689,21 @@ private void PopulateSessionProperties(OpenIdConnectMessage message, Authenticat
         protected virtual async Task<OpenIdConnectMessage> RedeemAuthorizationCodeAsync(OpenIdConnectMessage tokenEndpointRequest)
         {
             Logger.RedeemingCodeForTokens();
+
             var requestMessage = new HttpRequestMessage(HttpMethod.Post, _configuration.TokenEndpoint);
             requestMessage.Content = new FormUrlEncodedContent(tokenEndpointRequest.Parameters);
+
             var responseMessage = await Backchannel.SendAsync(requestMessage);
-            responseMessage.EnsureSuccessStatusCode();
-            var tokenResonse = await responseMessage.Content.ReadAsStringAsync();
-            var jsonTokenResponse = JObject.Parse(tokenResonse);
-            return new OpenIdConnectMessage(jsonTokenResponse);
+            var message = new OpenIdConnectMessage(await responseMessage.Content.ReadAsStringAsync());
+
+            if (string.IsNullOrEmpty(message.Error))
```

Comment:
This looks fragile and will likely cause a weird exception if the HTTP error response doesn't contain JSON (e.g because of a 500 response returned before the identity provider has a chance to send back a proper JSON error).